sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/implementation/directconnectivity/ReplicatedResourceClient.java

```diff
@@ -171,6 +167,81 @@ public Mono<StoreResponse> invokeAsync(RxDocumentServiceRequest request,
             addressSelector);
     }
 
+    private Mono<StoreResponse> getStoreResponseMonoWithSpeculation(RxDocumentServiceRequest request, Quadruple<Boolean, Boolean, Duration, Integer> forceRefreshAndTimeout) {
+        CosmosE2EOperationRetryPolicyConfig config = request.requestContext.getE2EOperationRetryPolicyConfig();
+        List<Mono<StoreResponse>> monoList = new ArrayList<>();
+        List<RxDocumentServiceRequest> requests = new ArrayList<>();
+
+        if (speculativeProcessor.getRegionsForPureExploration() != null
+            && !speculativeProcessor.getRegionsForPureExploration().isEmpty()) {
+            explore(request, forceRefreshAndTimeout);
+        }
+
+        if (speculativeProcessor.shouldIncludeOriginalRequestRegion()) {
+            monoList.add(getStoreResponseMono(request, forceRefreshAndTimeout));
+            requests.add(request);
+        }
+
+        for (URI endpoint : speculativeProcessor.getRegionsToSpeculate(config, this.transportClient
+            .getGlobalEndpointManager().getAvailableReadEndpoints())) {
+
+            if (request.requestContext.locationEndpointToRoute != endpoint) {
+                RxDocumentServiceRequest newRequest = request.clone();
+                newRequest.requestContext.routeToLocation(endpoint);
+                requests.add(newRequest);
+                monoList.add(getStoreResponseMono(newRequest, forceRefreshAndTimeout).delaySubscription(speculativeProcessor.getThreshold(config)));
```

Comment:
I am wondering how this will work for query for threshold based algo - the threshold is defined on operation, but for query, it will be per partition/feedRange request?  should the speculative processor be positioned in the RxDocumentClientImpl? (this is also where the e2e timeout is forced?)

Reply:
Also we would not want to speculate on all regions at the same time I guess? Wouldn't it make more sense to start after x ms on seocnd region, then aftera nother x ms on third region or similar?

Reply:
But yeah - for now I can also live with the current model

Reply:
Not very difficult to do. Can add another config parameter STEP that makes the speculation incremental

Reply:
Added threshold step